### PR TITLE
BLD: use unsigned instead of signed for lengths, avoid build warnings

### PR DESCRIPTION
**Pull request by jbrockmendel**
I'm much more confident about the other two BLD PRs (#26757, #26758) than this, largely because:

- [ ] in a couple of comments here I wrote that a certain term should always be non-negative.  I haven't actually confirmed that this is always the case
- [ ] This changed things from int64_t to uint64_t, but could also reasonably have changed them to size_t.  I don't have a compelling reason for the correctness of the uint64_t choice.

### CODE REVIEW

File: pandas/_libs/src/parser/tokenizer.c [hunk 1]
```diff
@@ -471,7 +472,7 @@ static int end_line(parser_t *self) {
         return 0;
     }
 
-    if (!(self->lines <= (int64_t) self->header_end + 1) &&
+    if (!(self->lines <= (uint64_t) self->header_end + 1) &&
```
Comment (WillAyd) [inline pandas/_libs/src/parser/tokenizer.c]: Is this cast necessary?

**Review by WillAyd — COMMENTED** [review]
On board with the concept just question around casts (applicable in a few places)

File: pandas/_libs/src/parser/tokenizer.c [hunk 2]
```diff
@@ -248,7 +248,8 @@ void parser_del(parser_t *self) {
 }
 
 static int make_stream_space(parser_t *self, size_t nbytes) {
-    int64_t i, cap, length;
+    uint64_t i, cap;
+    int64_t length;
```
Comment (WillAyd) [inline pandas/_libs/src/parser/tokenizer.c]: Does leaving this as is not introduce any new build warnings? Looks like assignments and comparisons are done with unsigned types within this function

File: pandas/_libs/src/parser/tokenizer.c (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/src/parser/tokenizer.c]: Good catch, I think it isn't.

File: pandas/_libs/src/parser/tokenizer.c [hunk 3]
```diff
@@ -651,7 +652,7 @@ static int parser_buffer_bytes(parser_t *self, size_t nbytes) {
     stream = self->stream + self->stream_len;                        \
     slen = self->stream_len;                                         \
     self->state = STATE;                                             \
-    if (line_limit > 0 && self->lines == start_lines + (int64_t)line_limit) {  \
+    if (line_limit > 0 && self->lines == start_lines + (uint64_t)line_limit) { \
```
Comment (WillAyd) [inline pandas/_libs/src/parser/tokenizer.c]: Is there a reason to even have these casts? Fraught with peril to begin with using macros I think this just adds nuances / developer burden

Comment (jbrockmendel) [reply pandas/_libs/src/parser/tokenizer.c]: Updated to remove these casts.

Should we consider getting rid of the macros?  That's outside my region of competence.

Comment (WillAyd) [reply pandas/_libs/src/parser/tokenizer.c]: Yea might be worth doing so in a separate PR. Here's the cpp take on macros:

https://isocpp.org/wiki/faq/inline-functions#inline-vs-macros

I could see this one being particular troublesome with our codebase:

https://isocpp.org/wiki/faq/misc-technical-issues#macros-with-if

I think the main downside to using inline vs macros would be support for pre-C99 compilers, but I don't think that's a problem given age and the fact that we have inline functions elsewhere in the code base

**Comment by WillAyd** [comment]
OK by me. Does this offer any performance improvements? Wondering if Cython optimizes for not doing a wraparound check like in this SO question:

https://stackoverflow.com/a/37933120/621736

File: pandas/_libs/src/parser/tokenizer.c (hunk 2 above)
Comment (jreback) [inline pandas/_libs/src/parser/tokenizer.c]: why is length signed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/src/parser/tokenizer.c (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/src/parser/tokenizer.c]: Because there weren't any warnings telling us not to.  Probably makes more sense unsigned liked the others.

Comment (jreback) [reply pandas/_libs/src/parser/tokenizer.c]: can you change this, otherwise lgtm.

**Comment by jreback** [comment]
small comment, merge master and ping on green.

File: pandas/_libs/src/parser/tokenizer.c (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/src/parser/tokenizer.c]: Changing this requires changing a few other places to match.  I think its benign, but not obvious.  Go for it?

Comment (jreback) [reply pandas/_libs/src/parser/tokenizer.c]: yep prob should

**Comment by jbrockmendel** [comment]
Ping

**Comment by WillAyd** [comment]
Thanks @jbrockmendel !

**Pull request merged**